### PR TITLE
make the service layer ignorant about the rest of the front-end

### PR DESCRIPTION
**Pull request by amdwit**
By removing imports from rest of front-end into service layer.
In order to 
prevent cyclic dependencies while mocking
make the code more maintainable and readable

### CODE REVIEW

File: app/frontend/src/service/api.ts [hunk 1]
```diff
@@ -5,7 +5,10 @@ import { ConversationsPromiseClient } from "../pb/conversations_grpc_web_pb";
 import { RequestsPromiseClient } from "../pb/requests_grpc_web_pb";
 import { SSOPromiseClient } from "../pb/sso_grpc_web_pb";
 
-import { store } from "../store";
+let getToken = () => "";
+export function setGetToken(newGetToken: () => string) {
+  getToken = newGetToken;
+}
```
Comment (lucaslcode) [inline app/frontend/src/service/api.ts]: This pattern used to be present, but @darrenvong managed to eliminate it using the `feature/x/index.ts` files to avoid circular import problems. 

Where were your circular imports coming from?

File: app/frontend/src/features/profile/index.ts [hunk 2]
```diff
@@ -2,10 +2,3 @@ import { ProtoToJsTypes } from "../../utils/types";
 import { UpdateProfileReq } from "../../pb/api_pb";
```
Comment (lucaslcode) [inline app/frontend/src/features/profile/index.ts]: These imports are unused

File: app/frontend/src/service/auth.ts [hunk 3]
```diff
@@ -1,35 +1,40 @@
-import client from "./api";
 import {
   LoginReq,
   SignupReq,
   SignupTokenInfoReq,
   UsernameValidReq,
 } from "../pb/auth_pb";
+import client from "./api";
+import { ServiceMap } from "./index";
```
Comment (lucaslcode) [inline app/frontend/src/service/auth.ts]: Unused import

File: app/frontend/src/service/search.ts [hunk 4]
```diff
@@ -1,18 +1,23 @@
-import client from "./api";
 import { SearchReq, User } from "../pb/api_pb";
+import client from "./api";
+import { ServiceMap } from "./index";
```
Comment (lucaslcode) [inline app/frontend/src/service/search.ts]: Unused import

File: app/frontend/src/service/user.ts [hunk 5]
```diff
@@ -1,140 +1,29 @@
...
-  const response = await client.api.getUser(
-    userReq,
-    token ? { authorization: `Bearer ${token}` } : undefined
-  );
-
-  return response.toObject();
-};
+import { ProtoToJsTypes } from "../utils/types";
+import client from "./api";
+import { ServiceMap } from "./index";
```
Comment (lucaslcode) [inline app/frontend/src/service/user.ts]: Unused import

File: app/frontend/src/index.tsx [hunk 6]
```diff
@@ -4,10 +4,13 @@ import { Provider } from "react-redux";
 import { PersistGate } from "redux-persist/integration/react";
 import "./index.css";
 import { persistor, store } from "./store";
+import * as api from "./service/api";
 
 import App from "./App";
 import * as serviceWorker from "./serviceWorker";
 
+api.setGetToken(() => store.getState().auth.authToken);
```
Comment (darrenvong) [inline app/frontend/src/index.tsx]: Why is this necessary? The app has just loaded at this point so there won't be an auth token anyway?

File: app/frontend/src/service/api.ts (hunk 1 above)
Comment (darrenvong) [reply app/frontend/src/service/api.ts]: Looks like it's introduced by a potentially unnecessary import that I don't really follow the need for in `index.tsx`. If we are using a global state management tool like Redux, we really should stick to it as the single source of truth for global states rather than trying to synchronise it in random module states, otherwise this becomes hard to reason about as the app grows.

Comment (amdwit) [reply app/frontend/src/service/api.ts]: Those circular imports were coming from the stories in story book, when trying to mock service methods.

Comment (amdwit) [reply app/frontend/src/service/api.ts]: btw: We are not using a single source of truth for global states as there is also some global state in the url. We use useParams to retrieve that state.

File: app/frontend/src/index.tsx (hunk 6 above)
Comment (amdwit) [reply app/frontend/src/index.tsx]: It is a function, so the api can always retrieve the latest version.

**Comment by amdwit** [comment]
I now decided to mock whole service modules instead of service methods in order to not have the cyclic dependency problem in test/storybook. I still prefer a service layer without dependencies back to other layers however.

File: app/frontend/src/service/api.ts (hunk 1 above)
Comment (darrenvong) [reply app/frontend/src/service/api.ts]: That's a good point, I'll backtrack a bit and should rephrase the single source of truth for a particular piece of state, which the old code broke by having it both in the Redux store and in the module and it's the main problem I have with it. Having multiple state sources isn't a problem (in fact, it's recommended by React like local component state) as long we rely on those to be the source of truth rather than sporadically syncing them all over the place.

Comment (amdwit) [reply app/frontend/src/service/api.ts]: But didn't my solution of setting a getter function keep a single source of truth? Anyway: I removed it. So I think you could approve the code.

**Review by lucaslcode — APPROVED** [review]
